homeassistant/components/alexa/smart_home.py

```diff
@@ -1276,6 +1354,152 @@ def async_api_previous(hass, config, request, entity):
+
+    data = {
+        ATTR_ENTITY_ID: entity.entity_id
+    }
+
+    payload = request[API_PAYLOAD]
+    if 'targetSetpoint' in payload:
+        temp = temperature_from_object(
+            payload['targetSetpoint'], unit)
+        if temp < min_temp or temp > max_temp:
+            return api_error_temp_range(
+                request, temp, min_temp, max_temp, unit)
+        data[ATTR_TEMPERATURE] = temp
+    if 'lowerSetpoint' in payload:
+        temp_low = temperature_from_object(
+            payload['lowerSetpoint'], unit)
+        if temp_low < min_temp or temp_low > max_temp:
+            return api_error_temp_range(
+                request, temp_low, min_temp, max_temp, unit)
+        data[climate.ATTR_TARGET_TEMP_LOW] = temp_low
+    if 'upperSetpoint' in payload:
+        temp_high = temperature_from_object(
+            payload['upperSetpoint'], unit)
+        if temp_high < min_temp or temp_high > max_temp:
+            return api_error_temp_range(
+                request, temp_high, min_temp, max_temp, unit)
+        data[climate.ATTR_TARGET_TEMP_HIGH] = temp_high
+
+    yield from hass.services.async_call(
+        entity.domain, climate.SERVICE_SET_TEMPERATURE, data, blocking=False)
+
+    return api_message(request)
+
+
+@HANDLERS.register(('Alexa.ThermostatController', 'AdjustTargetTemperature'))
+@extract_entity
+@asyncio.coroutine
+def async_api_adjust_target_temp(hass, config, request, entity):
+    """Process an adjust target temperature request."""
+    unit = entity.attributes[CONF_UNIT_OF_MEASUREMENT]
+    min_temp = entity.attributes.get(climate.ATTR_MIN_TEMP)
+    max_temp = entity.attributes.get(climate.ATTR_MAX_TEMP)
+
+    temp_delta = temperature_from_object(
+        request[API_PAYLOAD]['targetSetpointDelta'], unit, interval=True)
+    target_temp = float(entity.attributes.get(ATTR_TEMPERATURE)) + temp_delta
+
+    if target_temp < min_temp or target_temp > max_temp:
+        return api_error_temp_range(
+            request, target_temp, min_temp, max_temp, unit)
+
+    data = {
+        ATTR_ENTITY_ID: entity.entity_id,
+        ATTR_TEMPERATURE: target_temp,
+    }
+
+    yield from hass.services.async_call(
+        entity.domain, climate.SERVICE_SET_TEMPERATURE, data, blocking=False)
+
+    return api_message(request)
+
+
+@HANDLERS.register(('Alexa.ThermostatController', 'SetThermostatMode'))
+@extract_entity
+@asyncio.coroutine
+def async_api_set_thermostat_mode(hass, config, request, entity):
+    """Process a set thermostat mode request."""
+    mode = request[API_PAYLOAD]['thermostatMode']
+
+    operation_list = entity.attributes.get(climate.ATTR_OPERATION_LIST)
+    # pylint: disable=stop-iteration-return
+    ha_mode = next(
+        (k for k, v in API_THERMOSTAT_MODES.items() if v == mode),
+        None
+    )
+    if ha_mode not in operation_list:
+        msg = 'The requested thermostat mode {} is not supported'.format(mode)
+        _LOGGER.info(msg)
```

Comment:
This should be an error ? And should we have the alexa smart home handler just log an error when we are returning an error message?

Reply:
Totally valid requests from the user can specify modes or temperatures not supported by the device (since the API does not enumerate the accepted values) and I'm logging them at `info` level since they're not indicative of software bugs.

You're right, it might make sense to log all error responses we send though.